[Bug 14630] com.livecode.list: Add "offset" and "index" syntax for lists

peter-b opened the pull request:
Add a number of new list manipulation syntax definitions and corresponding implementations.

Finding elements in lists:

```
index of _ in _
first index of _ in _
last index of _ in _
index of _ before _ in _
first index of _ before _ in _
last index of _ before _ in _
index of _ after _ in _
first index of _ after _ in _
last index of _ after _ in _
```

Finding subsequences in lists:

```
offset of _ in _
first offset of _ in _
last offset of _ in _
offset of _ before _ in _
first offset of _ before _ in _
last offset of _ before _ in _
offset of _ after _ in _
first offset of _ after _ in _
last offset of _ after _ in _
```

Since strings/data are self-flattening sequences of chars/bytes, the `index` operations are added to strings and data, identical to the corresponding `offset` operations.  The `offset of chars` and `offset of bytes` syntax is removed; it's the default chunk type for and should be implicit.

In order to make `index` available as a syntax keyword, the deprecated `index` and `uindex` types are removed, and `LCIndex` and `LCUindex` should be used instead.

runrevmark commented:
(Replicated from BZ to make it more visible)

I'm not sure I'm 100% happy with 'the offset of list' - as it is slightly confusing.

Could we perhaps consider using 'index' and 'offset' as the two distinct operations?

INDEX

the index of [ <chunk> ] tValue in tSequence:
  returns the first (1-based) index of tValue in tSequence

Here tSequence is considered to be a sequence of <chunk> and tValue should be compatible with <chunk>. It returns 0 if tValue is not found in tSequence.

OFFSET

the offset of [ <chunk> ] tValue in tSequence:
  returns where the first instance of tValue starts in tSequence

Here tSequence is considered to be a sequence of <chunk> - it does a subsequence match.

So:

the index of [ char ] <character> in <string>
the index of [ byte ] <byte> in <data>
the index of [ element ] <any value> in <list>

And:

the offset of [ chars ] <string> in <string>
the offset of [ bytes ] <data> in <data>
the offset of [ elements ] <list> in <list>

I think this captures the difference between element-wise match and substring match. For strings 'offset' and 'index' are almost the same thing (strings being self-flattening sequences of chars, chars being strings ;)) - except that we can be much more strict with 'index'.

runrevmark commented:
Indeed, both offset and list are target type driven - so I wonder if 'char', 'byte' and 'element' are needed at all:

If the haystack is a string, then the needle must be a char (string)
If the haystack is a data, then the needle must be a byte (data)
If the haystack is a list, then needle can be any value (index) or must be a list (offset)

I think we might have got slightly muddled up with the string-chunk index/offset operation - where you _do_ need the clarifying 'chunk' type as chunks (items, lines etc.) are all strings.

runrevmark commented:
@peter-b: Semantic / syntax decisions aside - I think I've figured out the problem with this branch... When it merges the two syntax rules, its not remarking the 'marks' correctly. Upshot: the list of applicable methods for the syntax at a call-site is empty for the provided parameters - hence 'non-evaluatable'.

peter-b commented:
@runrevmark I think the `index` and `offset` distinctions are clear.  I'm not sure what the exact implications for the `string`/`char` syntax are though.

I suggest that I should implement `index` for lists, and then figure out from that what the semantics should be for strings...?

runrevmark commented:
@peter-b: Yes, so to clarify...

I think instead of 'the offset of list x in y' should be just:

```
  the offset of x in y
```

And what we currently have as 'the offset of element x in y' should be:

```
  the index of x in y
```

Additionally, I don't think the optional modifiers are necessary as the type of index/offset operation can be determined by the type of the target.

Adding string char and byte 'index' operations is perhaps a separate thing :)

peter-b commented:
Okay. What about removing "offset of chars" variants from strings?

runrevmark commented:
The 'offset of chars' variant isn't needed, and 'offset of bytes' should be just 'offset of' I think.

peter-b committed:
com.livecode.byte: Remove "offset of bytes" syntax.

Just use the generic "offset of" syntax instead.

peter-b committed:
com.livecode.char: Add some extra tests for offset of empty string.

peter-b committed:
lc-compile: Correctly handle deduplicated syntax nodes.

When merging syntax, the compiler replaces deduplicated rules with
NULL in the syntax tree.  When printing the syntax tree out, these
removed rules need to be skipped in order to avoid a NULL-dereference
crash.

peter-b committed:
com.livecode.foreign: Remove `index` and `uindex` foreign types.

The `LCIndex` and `LCUIndex` types should be used instead.

`index` will be repurposed as a syntax keyword.

peter-b committed:
com.livecode.char: Remove "offset of chars" syntax variant.

Only provide generic "offset of".

peter-b committed:
com.livecode.list: Add "index of _ after _ in _" accessor.

Add a new operation on lists:

    the index of <tNeedle> after <tPosition> in <tList>

This looks for an element equal to <tNeedle> in the subsequence of
<tList> starting at the 1-based index <tPosition> (exclusive).  It
returns the 1-based index relative to <tPosition>, or 0 if no element
of of the subsequence is equal to <tNeedle>.

There are two variants of the operation:

    the first index of <tNeedle> after <tPosition> in <tList>
    the last index of <tNeedle> after <tPosition> in <tList>

If <tNeedle> is a list, it is considered as an element of <tList> not
as a subsequence of <tList>.

peter-b committed:
com.livecode.char: Corrections to offset tests.

peter-b committed:
libfoundation: Add MCProperListFirstIndexOfElementInRange().

Rework `MCProperListFirstIndexOfElement()` to redirect onto the new
function.

peter-b committed:
libfoundation: Add MCProperListLastIndexOfElementInRange().

peter-b committed:
com.livecode.list: Add "index of _ in _" accessor.

Add a new operation on lists:

    the index of <tNeedle> in <tList>

which returns the 1-based index of the first element in <tList> that
is equal to <tNeedle>, or 0 if no element of <tList> is equal to
<tNeedle>.

There are two variants of the operation:

    the first index of <tNeedle> in <tList>
    the last index of <tNeedle> in <tList>

If <tNeedle> is a list, it is considered as an element of <tList> not
as a subsequence of <tList>.

peter-b committed:
com.livecode.list: Add "index of _ before _ in _" accessor.

Add a new operation on lists:

    the index of <tNeedle> before <tPosition> in <tList>

This looks for an element equal to <tNeedle> in the subsequence of
<tList> ending at the 1-based index <tPosition> (exclusive).  It
returns the 1-based index, or 0 if no element of of the subsequence is
equal to <tNeedle>.

There are two variants of the operation:

    the first index of <tNeedle> before <tPosition> in <tList>
    the last index of <tNeedle> before <tPosition> in <tList>

If <tNeedle> is a list, it is considered as an element of <tList> not
as a subsequence of <tList>.

peter-b committed:
com.livecode.char: "index" and "offset" are the same operation for strings.

Strings are self-flattening sequences of graphemes.

peter-b committed:
com.livecode.list: Add "offset of _ in _" accessor.

    
Add a new operation on lists:
    
    the offset of <tNeedle> in <tList>
    
which returns the 1-based index of the start of the first subsequence
in <tList> that is equal to <tNeedle>, or 0 if no subsequence of
<tList> is equal to <tNeedle>.
    
There are two variants of the operation:
    
    the first offset of <tNeedle> in <tList>
    the last offset of <tNeedle> in <tList>
    
<tNeedle> must be a List.

peter-b committed:
libfoundation: Rename MCProperListFirstIndexOfList().

For consistency with the string operation naming, rename
`MCProperListFirstIndexOfList()` to `MCProperListFirstOffsetOfList()`.

peter-b committed:
libfoundation: Add MCProperListLastOffsetOfListInRange().

peter-b committed:
libfoundation: Add MCProperListFirstOffsetOfListInRange().

Add a new function that finds the first offset of the start of a list
in another list.

peter-b committed:
com.livecode.list: Add "offset of _ after _ in _" accessor.

Add a new operation on lists:

   the offset of <tNeedle> after <tPosition> in <tList>

This looks for an subsequence equal to <tNeedle> in the subsequence of
<tList> starting at the 1-based index <tPosition> (exclusive).  It
returns the 1-based index relative to <tPosition> of the first element
of the matching subsequence, or 0 if no matching subsequence is found.

There are two variants of the operation:

    the first offset of <tNeedle> after <tPosition> in <tList>
    the last offset of <tNeedle> after <tPosition> in <tList>

<tNeedle> must be a List.

peter-b committed:
com.livecode.list: Add "offset of _ before _ in _" accessor.
    
Add a new operation on lists:
    
   the offset of <tNeedle> before <tPosition> in <tList>
    
This looks for an subsequence equal to <tNeedle> in the subsequence of
<tList> ending at the 1-based index <tPosition> (exclusive).  It
returns the 1-based index in <tList> of the first element of the
matching subsequence, or 0 if no matching subsequence is found.
    
There are two variants of the operation:
    
    the first offset of <tNeedle> before <tPosition> in <tList>
    the last offset of <tNeedle> before <tPosition> in <tList>
    
<tNeedle> must be a List.

peter-b commented:
@runrevmark I think that should be sufficiently comprehensive.

peter-b commented:
Apparently not; I need to add `index` for `Data`.

peter-b committed:
com.livecode.byte: "index" and "offset" are the same operation for data.

livecodefraser commented:
Everyone happy with this now?

peter-b commented:
@livecodefraser I'm happy if @runrevmark is. :grinning:

peter-b committed:
[Bug 14846] com.livecode.list: Disable "(index|offset) of _ (before|after) _"

peter-b committed:
[Bug 14846] com.livecode.byte: Disable "(index|offset) of _ (before|after) _"

peter-b committed:
[Bug 14846] com.livecode.char: Disable "(index|offset) of _ (before|after) _"

runrevmark commented:
Reviewed-by: @runrevmark.
